src/libstore/aws-creds.cc

```diff
@@ -0,0 +1,246 @@
+#include "nix/store/aws-creds.hh"
+
+#if NIX_WITH_S3_SUPPORT
+
+#  include "nix/util/finally.hh"
+#  include "nix/util/logging.hh"
+#  include "nix/util/sync.hh"
+#  include "nix/util/util.hh"
+
+#  include <aws/crt/Api.h>
+#  include <aws/crt/auth/Credentials.h>
+#  include <aws/crt/io/Bootstrap.h>
+
+#  include <condition_variable>
+#  include <map>
+#  include <mutex>
+#  include <unistd.h>
+
+namespace nix {
+
+namespace {
+
+// AWS CRT initialization
+static bool initAwsCrt()
+{
+    static bool initialized = []() {
+        try {
+            // Use a static local variable instead of global to control destruction order
+            struct CrtWrapper
+            {
+                Aws::Crt::ApiHandle apiHandle;
+                CrtWrapper()
+                {
+                    apiHandle.InitializeLogging(Aws::Crt::LogLevel::Warn, (FILE *) nullptr);
+                }
+                ~CrtWrapper()
+                {
+                    try {
+                        // CRITICAL: Clear credential provider cache BEFORE AWS CRT shuts down
+                        // This ensures all providers (which hold references to ClientBootstrap)
+                        // are destroyed while AWS CRT is still valid
+                        clearAwsCredentialsCache();
+                        // Now it's safe for ApiHandle destructor to run
+                    } catch (...) {
+                        ignoreExceptionInDestructor();
+                    }
+                }
+            };
+            static CrtWrapper crt;
+            return true;
+        } catch (const std::exception & e) {
+            debug("Failed to initialize AWS CRT: %s", e.what());
+            return false;
+        } catch (...) {
+            debug("Failed to initialize AWS CRT: unknown error");
+            return false;
+        }
+    }();
+    return initialized;
+}
+
+// Internal credential provider class
+class AwsCredentialProvider
+{
+private:
```

Comment:
Now that this is private to this file (in no header), I don't think we need this class. We can just have free `static` functions on `std::shared_ptr<Aws::Crt::Auth::ICredentialsProvider>` directly.

Reply:
(`static` in the sense of private to this compilation unit, not in the C++ `struct` sense)